src/pip/_internal/download.py

```diff
@@ -140,13 +145,31 @@ def __init__(self, prompting=True):
         self.passwords = {}
 
     def __call__(self, req):
-        parsed = urllib_parse.urlparse(req.url)
+        # Get credentials for this request
+        url, username, password = self._get_url_and_credentials(req.url)
+
+        # Set the url of the request to the url without any credentials
+        req.url = url
+
+        if username or password:
+            # Send the basic auth with this request
+            req = HTTPBasicAuth(username or "", password or "")(req)
+
+        # Attach a hook to handle 401 responses
+        req.register_hook("response", self.handle_401)
+
+        return req
+
+    def _get_url_and_credentials(self, original_url,
+                                 allow_netrc=True, allow_keyring=True):
+
+        parsed = urllib_parse.urlparse(original_url)
 
         # Get the netloc without any embedded credentials
         netloc = parsed.netloc.rsplit("@", 1)[-1]
 
-        # Set the url of the request to the url without any credentials
-        req.url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
+        # Return the url without any credentials
+        url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
```

Comment:
Would `remove_auth_from_url()` work for this purpose?
https://github.com/pypa/pip/blob/ebfa66361e92b3cac7c259956effcd55617c1486/src/pip/_internal/utils/misc.py#L918-L922
It has a number of unit tests, which is helpful.

Reply:
It would, but we'd still need to parse the URL to get the netloc returned, so it doesn't really gain much (unless we want a new util function to remove and return auth from a URL? But that doesn't compose so easily from what's already there)